### ISSUE TITLE
Smoothing function for surface data

### ISSUE THREAD

**Issue by NicolasGensollen** [nilearn/nilearn#2949]
See this [Neurostars post](https://neurostars.org/t/smoothing-surface-data/20159) and #2747 for a related issue.

The idea would be to have a smoothing function for surface data similar to `nilearn.image.smooth_img` for volumes:

https://github.com/nilearn/nilearn/blob/cc99fa76df137e9ac6010c311e15fa9ac12ce66d/nilearn/image/image.py#L244-L248

After discussing this, we agreed that this could be a useful addition, and that it would fit within the scope of Nilearn (most likely within the surface module).

One point that still needs to be discussed is how this function should be implemented as there are several possible ways to do that. 

For example:

For each point:
- Average across mesh direct neighbors
- Average across nodes within a ball of a given radius
- Average across mesh neighborhood (all nodes within some graph distance defined through edge weights)
- Other ideas?

How should we average?

Can we preserve the same API as `smooth_img` (with `fwhm` as input)?

_Note: Some of these implementations are obviously not desirable as they can introduce biases in the smoothing._ 
_Note: The proposed implementation should be fast (avoid for loops...) and not overly complex for maintenance reasons._

**Comment by GaelVaroquaux** [comment]
Weighted average of the neighbors, with the weight being function of the distance between the two nodes (approximates a spatial kernel on the mesh). More smoothing would be achieved by iterating this procedure to propagate information further.

**Comment by jyeatman** [comment]
Agree with @GaelVaroquaux with one clarification: by distance you mean distance along the mesh not distance in 3D coordinates right? In other words, the smoothing will always move along the faces of the mesh as oppose to volume smoothing which just goes based on distance in the volume. We want to make sure we preserve the cortical topography and don't (for example) jump a sulcus with the smoothing

**Comment by bthirion** [comment]
+1 for mesh-based distances.
It is easy to extract the k-topological neighbors of a vertex and to compute distances along the mesh on those.

**Comment by noahbenson** [comment]
Hi there! I'm the author of [neuropythy](https://github.com/noahbenson/neuropythy)—just wanted to drop a note in this thread that neuropythy supports a lot of these surface operations mentioned here. I'm happy to help with implementing any of this if it's needed (or to help scrape the relevant code out of neuropythy).

Neuropythy is made to work with surfaces primarily, and it doesn't currently integrate with `nilearn` (though building out this integration is definitely a long-term goal). I've included some sample code that demonstrates how these surface operations would work. There's potentially a lot more to discuss here, but for now I'll just say that I'm interested in helping.

(One caveat is that neuropythy is currently licensed under the Affero GPL so it may not be directly compatible with nilearn. I'm happy to relicense any relevant code to nilearn in this case. Also, part of the goal of building out neuropythy's integrations with other libraries like nilearn is to break the library down into smaller libraries with less restrictive licenses, so this is somewhat temporary.)

* **Average a surface property using neighboring vertices**. This example is just for the LH:
  ```python
  import neuropythy as ny, numpy as np
  sub = ny.freesurfer_subject('/path/to/a/subject')
  prop = sub.lh.prop('thickness') # or load in a property for the LH
  neighborhoods = sub.lh.tess.neighborhoods
  avg_prop = [np.mean(prop[nn]) for nn in neighborhoods]
  ```
  One could also turn `neighborhoods` into a sparse matrix `M` such that multiplying by a property vector yields the distance-weighted average (as proposed by @GaelVaroquaux), then `(M**k).dot(prop)` would give you the `k`th iteration of the smoothing.
* **Calculate the all-pairs distances on a (sub-) mesh.** This example uses just the LH V1 ROI since calculating the all-pairs distances for the entire mesh requires a matrix whose size is something like 150,000 by 150,000. To do whole-mesh surface smoothing using a Gaussian kernel via this method, one would probably need to do something clever like break the problem down into many submeshes.
   ```python
   mesh = sub.lh.surface('midgray')
   v1_mask = mesh.mask('V1_label') # V1 label is from FreeSurfer's recon-all
   v1_submesh = mesh.submesh(v1_mask)
   all_pairs_dist = v1_submesh.dijstra()
   ```

**Comment by bthirion** [comment]
Thk you for the comment. Solution 1, with a sparse matrix, is indeed a relevant.
It should be combined with solution 2 to derive geodesic distance, but restricted to a local  neighborhood of each vertex (and thus also, with a sparse representation that makes it memory-efficient).

**Comment by alexisthual** [comment]
I previously used [`gdist`](https://github.com/the-virtual-brain/tvb-gdist) to compute matrices of geodesic distances on meshes in python and thought it worked well (although I don't know if it scales to 100k-vertex meshes) :slightly_smiling_face:.
I don't know if it makes sense to add this dependency to nilearn, but I guess it could help us to easily derive an approximation of the distance kernel using [`gdist.compute_gdist()`](https://github.com/the-virtual-brain/tvb-gdist#apis):

```python
import gdist
import numpy as np
from sklearn.kernel_approximation import Nystroem

# load vertices coordinates and faces for given mesh
vertices, triangles = ...

# wrap gdist.compute_gdist to use it as a metric
def metric(source_index, target_index):
    return gdist.compute_gdist(
        vertices, triangles,
        np.array([source_index]), np.array([target_index])
    )

# compute kernel matrix approximation
nys = Nystroem(metric=metric, n_components=100)
embedding = nys.fit_transform(np.arange(vertices.shape[0]))

# load sparse neighbouring matrix (assert it is a boolean matrix)
connectivity = ...

# load a surface map on the mesh (could be an fmri contrast for instance)
surface_map = ...

# smooth surface_map using up-to-p-voxels-away neighbours
neighbours = (connectivity ** p).indices
smoothed_surface_map = np.sum(
    surface_map[neighbours] * np.linalg.norm(embedding[neighbours] - embedding[:, None], ord=2, axis=2),
    axis=2
)
```

I didn't test any of this, so maybe it's all wrong, but I thought it might work!

**Comment by bthirion** [comment]
I wouldn't have gdist as a nilearn dependency. Thinking of it: if the smoothing kernels are local, then we only need to have the geodedic distances close to each vertex. In that case the Euclidean distance is OK I think.

**Comment by bthirion** [comment]
You can open a PR whenever you want. This will make interactions easier.

**Comment by clane9** [comment]
Hi there, I see that a PR is in progress. But I thought it would still be worth bringing up a slightly different smoothing approach. In the HCP pipeline ([paper](https://surfer.nmr.mgh.harvard.edu/ftp/articles/2013/2013_-_Glasser_et_al._-_NeuroImage.pdf) p 121, [docs](https://www.humanconnectome.org/software/workbench-command/-metric-smoothing)), they basically do a Gaussian weighted averaging of all vertices within a 3 sigma geodesic radius.

I like that this approach is analogous to volume smoothing. And it should still be fairly efficient to compute. Any thoughts?

**Comment by noahbenson** [comment]
I'm not sure that calculating this is easy, but I'd be glad to be proven wrong. The issue as I see it is that you'd need to approximate the geodesic with Dijkstra's method (alternately, something much more complicated), and the algorithm for that in (e.g.) SciPy runs over the entire mesh/graph without a good way to limit it to a maximum distance to 3σ last I checked. 

In theory, one could calculate the all-pairs shortest path between all vertices, stopping the search from a particular vertex after you've reached a distance of 3σ from it and store the results in a sparse matrix. Without the stop condition (or without a _sparse_ matrix), the matrix will be too big to fit in memory on most systems. Unfortunately, I don't know of an efficient Python implementation of this off-hand (maybe there's an implementation of this in a standard library that I don't know about?). Any pure Python implementation would be very slow because it would involve a lot of looping over vertices, and I'm not sure if nilearn wants to support a C implementation? The HCP WorkBench code I believe is in C, so they probably have an implementation if someone wants to try to adapt it.

Maybe I'm missing something?

**Comment by clane9** [comment]
This is a very fair point, thanks @noahbenson! In principle, it seems like it should be possible to do it fairly efficiently as you describe. Basically running a Dijkstra shortest path search with early stopping. But whether this can be implemented easily without resorting to a C extension is another question..

It looks like [`scipy.sparse.csgraph.dijkstra`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csgraph.dijkstra.html) *almost* does what we would want. It has a distance `limit` argument for stopping the search. However the returned distance matrix is still dense, with above threshold values defaulting to inf. (See also the [source](https://github.com/scipy/scipy/blob/372579fcb6085f298e275c568f2f89fb36a07592/scipy/sparse/csgraph/_shortest_path.pyx#L821).)

I wrote an initial implementation of a sparse geodesic radius search using just NumPy/SciPy and ran a comparison with a few other search methods. See [this notebook gist](https://gist.github.com/clane9/7f5ddf65756a9ccf84dd1e9fa3ba063f). This is the take-away:

![image](https://user-images.githubusercontent.com/11963692/175353463-dab578c9-533e-4abb-af19-f9dcb29c788a.png)

The proposed algorithm ("sparse-graph") is definitely slower than `scipy.sparse.csgraph.dijkstra` ("dijkstra"), scikit-learn `NearestNeighbors` ("nn-search"), or brute-force search. E.g. it's 4x slower than dijkstra at 25k points. So perhaps you are right that computing geodesic radius neighborhoods in pure python is too expensive.

**Comment by bthirion** [comment]
My impression is that it could be good enough.
